compiler/codegen/cg-symbol.cpp

```diff
@@ -2572,7 +2572,8 @@ void FnSymbol::codegenPrototype() {
 
     if (generatingGPUKernel) {
       func->setConvergent();
-      if (!hasFlag(FLAG_GPU_AND_CPU_CODEGEN)) {
+      if (!hasFlag(FLAG_GPU_AND_CPU_CODEGEN) &&
+          !hasFlag(FLAG_GPU_SPECIALIZATION)) {
```

Comment:
Scrolling down, I see that you added this flag to a stub function as well. I believe this check and that addition are related because you wanted the stub function to not have a kernel calling convention. I find this reliance on `FLAG_GPU_SPECIALIZATION` a bit tangly, because that flag was added for a different reason:

`--gpu-specialization` flag enables creating function clones downstream from any `on` statement where the assumption is if that `on` statement is targeting a GPU, GPU-eligible loops wouldn't require per-loop checks to see whether you are on a GPU sublocale or not. You are absolutely right that "GPU specialization" sounds like what you are doing here, but there are 2 different kinds of specialization here: this function (a) executes on a GPU sublocale, (b) called on device. (a) is a superset of (b) and what `FLAG_GPU_SPECIALIZATION` means in the rest of the code. (b) is what you need to be more precise. https://github.com/Cray/chapel-private/issues/5926#issuecomment-1930684763 has more detailed thoughts on this)

Back to the problem at hand, I think we are using these flags a bit randomly already. But something like this makes sense to me and I believe requires the least amount of change in the compiler:

- Compiler generated kernel: `GPU_CODEGEN` and `GPU_KERNEL` (a new flag)
- Compiler generated GPU _device_ specialization: `GPU_CODEGEN` (and definitely no `GPU_CPU_CODEGEN`)
- Non-specialized function that can be called from the device and the host: `GPU_CPU_CODEGEN` (that should be the case today) 
- Function generated by `--gpu-specialization`: `GPU_SPECIALIZATION`

What do you think?